### ISSUE TITLE
Configurable defaults for `ipfs add`

### ISSUE THREAD

**Issue by swedneck** [ipfs/go-ipfs#7749]
Currently you have to manually specify non-default settings using commandline flags (e.g. `ipfs add --raw-leaves -s buzhash`).   
It would be useful to be able to change these defaults in the config.

**Comment by aschmahmann** [comment]
I get the idea here, but having consistency and copy-pasteablility of CLI commands seems more valuable.

e.g. if a user `ipfs add ubuntu.iso` is causing problems I would have no way of knowing what they did. This also doesn't seem like a common CLI pattern, e.g. I almost always do `ls -al` but without me writing a custom alias I always have to pass `-al`.

**Comment by swedneck** [comment]
Can you clarify what the problem with this idea is? 
I'm not saying to remove CLI flags, rather to make the *defaults* configurable instead of hardcoded.

**Comment by aschmahmann** [comment]
There's nothing inherently wrong with the idea, and having configurable defaults for certain things is something we may end up doing (e.g. if we version the API and want to allow people to allow their scripts to use an old API).

However, by hiding the parameters actually used in a command there's some reproducibility that's lost. For example, perhaps you set the default to buzhash last year and you're getting a panic and so you file an issue `ipfs add` causes a panic. While you should probably dump your config file into the issue it's super easy to miss that you're using a different set of options.

Basically my thought is that this isn't inherently problematic but we should be careful how frequently we do this since the more options we hide in the config file the more complex the config file gets and the more context is needed to understand what you're doing.

**Comment by swedneck** [comment]
That makes sense.

Just spit-balling here, but what if you had setting presets?  
For example you could have a custom preset called "foobar" which specifies the flags `--cid-version=1 --nocopy -w`, you could then use this like `ipfs add --preset=foobar` to override the defaults with the flags in the preset. Ideally any further flags supplied would work like normal, including overriding the preset.

**Comment by aschmahmann** [comment]
Is your proposal to in the config file let users define `ipfs add --preset=foobar` to utilize some flags, or that go-ipfs should come with these presets bundled in?

Both aren't particularly appealing to me (although I may be alone here 🤷). Setting it in the config could work but seems a bit like overkill to start adding those in especially since you could just write shell scripts that do that for you. Having the presets built in feels awkward because the short names aren't very descriptive and just lead to people looking around in documentation to figure out what they mean (e.g. our `ipfs init` profiles).

Basically, it's not clear to me that adding this feature in is worthwhile given that a script could emulate the behavior and it adds complexity to the already long config file. Are there any CLI applications where you've made use of this successfully in the past? 

----

Note that with go-ipfs the CLI matches the HTTP API which adds a whole other layer of issues or work here. For example, the go-ipfs configs are defined on the daemon. This means that presets would be defined on a "server" and exposed via an HTTP API when the config options should really be on the client.

**Comment by lidel** [comment]
Addressed by `Import.*` config, added in [v0.29](https://github.com/ipfs/kubo/releases/tag/v0.29.0) and expanded through [v0.40](https://github.com/ipfs/kubo/releases/tag/v0.40.0):

- `Import.CidVersion`, `Import.UnixFSRawLeaves`, `Import.UnixFSChunker`, `Import.HashFunction` (v0.29)
- `Import.UnixFSFileMaxLinks`, `Import.UnixFSDirectoryMaxLinks`, `Import.UnixFSHAMTDirectoryMaxFanout` (v0.35)
- `Import.UnixFSDAGLayout`, `Import.UnixFSHAMTDirectorySizeEstimation` (v0.40)
- Config profiles `unixfs-v1-2025` and `unixfs-v0-2015` for one-command presets (v0.40)
